Implemented Maybe assignment

LightnessOfBeing opened the pull request.

LightnessOfBeing committed:
Implemented Maybe assignment

LightnessOfBeing committed:
Deleted unnecessary files.

LightnessOfBeing committed:
Replaced assertEquals with assertFalse and assertTrue.

yurii-litvinov commented on Maybe/src/main/java/Maybe.java:
```diff
@@ -0,0 +1,96 @@
+ * Exception class for Maybe.
+ */
+
+class MyException extends Exception{
```
Класс вне пакета, плохо

yurii-litvinov replied:
И надо пробел перед "{", здесь и ниже

yurii-litvinov replied:
Исключения в подавляющем большинстве случаев делают public, иначе извне пакета такое исключение не поймать

yurii-litvinov commented on Maybe/src/main/java/Maybe.java:
```diff
@@ -0,0 +1,96 @@
+ * @param <T> - represents type.
+ */
+
+public class Maybe<T> {
```
Лучше делать каждый класс в своём файле --- Java-way и всё такое

yurii-litvinov commented on Maybe/src/main/java/Maybe.java:
```diff
@@ -0,0 +1,96 @@
+     */
+
+    public static <T> Maybe<T> nothing(){
+        return new Maybe<>(null);
```
Лучше статическое поле, чтобы тысячи одинаковых объектов не создавать (и заодно они все равны будут)

yurii-litvinov commented on Maybe/src/main/java/Maybe.java:
```diff
@@ -0,0 +1,96 @@
+    /**
+     *
+     * @return value that is storing in instance
+     * @throws MyException
```
Надо всё-таки комментарий написать

yurii-litvinov commented on Maybe/src/test/java/MaybeTest.java:
```diff
@@ -0,0 +1,145 @@
+        File file = new File(directory + "/file.in");
+        Scanner s = new Scanner(file);
+
+        assertTrue(MaybeTest.readNumber(s).get().equals(1));
```
А чего не assertEquals?

yurii-litvinov commented on Maybe/src/test/java/MaybeTest.java:
```diff
@@ -0,0 +1,145 @@
+        assertFalse(MaybeTest.readNumber(s).isPresent());
+        assertFalse(MaybeTest.readNumber(s).isPresent());
+
+        s.close();
```
try-with-resources

yurii-litvinov commented on Maybe/src/test/java/MaybeTest.java:
```diff
@@ -0,0 +1,145 @@
+            e.printStackTrace();
+        }
+
+        sIn.close();
```
Это тоже лучше с помощью try-with-resources закрывать

yurii-litvinov commented on Maybe/src/test/java/MaybeTest.java:
```diff
@@ -0,0 +1,145 @@
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
Ловить исключения в тестах нельзя

yurii-litvinov commented on Maybe/src/test/java/MaybeTest.java:
```diff
@@ -0,0 +1,145 @@
+        try {
+            assertTrue(m.map(x -> x * x).get().equals(4));
+        } catch(MyException e) {
+            fail(e.getMessage());
```
Это сам jUnit делает, не надо ловить исключения в тестах

yurii-litvinov reviewed (changes requested):
Вообще, функциональность по возведению в квадрат и записи в файл ожидалась в основном коде программы, а не в тестах. Сделайте, пожалуйста, ещё один класс.

LightnessOfBeing committed:
Fixed all complaints.

LightnessOfBeing committed:
Added Main class, where located additional methods.

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,53 @@
+import static java.lang.Integer.parseInt;
+
+@SuppressWarnings({"ConstantConditions", "ResultOfMethodCallIgnored"})
+public class Main {
```
Раз в нём делается какая-то нетривиальная работа, то ему тоже нужен комментарий

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,53 @@
+public class Main {
+
+    @Nullable
+    public static Maybe<Integer> readNumber(@NotNull Scanner s) {
```
И паблик-методам, как обычно, нужны комментарии

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,53 @@
+        Maybe<Integer> m;
+        output.createNewFile();
+        ArrayList<String> written = new ArrayList<>();
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(path2)); Scanner sIn = new Scanner(input)) {
```
Лучше каждое объявление на своей строке, и использовать более говорящие имена, чем sIn

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @NotNull
+    public static ArrayList<String> readAndSquare(@NotNull String path1, @NotNull String path2) throws IOException, MyException {
```
path1 и path2 --- имена не позволяют быстро догадаться, кто из них кто

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,53 @@
+                }
+            }
+        }
+        output.delete();
```
ммм?

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/MyException.java:
```diff
@@ -0,0 +1,17 @@
+ * Exception class for Maybe.
+ */
+
+public class MyException extends Exception {
```
Ну не то чтобы хорошее название для исключения :)

yurii-litvinov commented on Maybe/src/test/java/ru/spbau/mit/vishnyakov/MaybeTest.java:
```diff
@@ -0,0 +1,90 @@
+        ArrayList<String> actual = Main.readAndSquare(sourcePath, destinationPath);
+
+        String[] answer = {"1", "4", "9", "16", "25", "10000", "null", "null"};
+        assertArrayEquals(answer, actual.toArray());
```
А лучше ещё реально открыть файл с результатом и прочитать оттуда. В условии ведь было "записать файл", надо проверить, что условие выполняется

LightnessOfBeing committed:
Added commentaries to Main class. Renamed exception class, added proper testing.

yurii-litvinov commented on Maybe/src/main/java/ru/spbau/mit/vishnyakov/Main.java:
```diff
@@ -0,0 +1,79 @@
+     * Reads a file and saves squared number to another file.
+     * @param sourcePath where input file is located.
+     * @param destinationPath where the result file will be located.
+     * @return
```
Пустые тэги не нужны

yurii-litvinov reviewed (approved):
Ага, зачтена
